lib/sidekiq_queue_metrics/configuration.rb

```diff
@@ -1,11 +1,26 @@
 module Sidekiq::QueueMetrics
+  def self.support_death_handlers?
+    Sidekiq::VERSION >= '5.1'
+  end
+
   def self.init(config)
     config.server_middleware do |chain|
       chain.add Sidekiq::QueueMetrics::JobSuccessMonitor
     end
 
-    config.death_handlers << Sidekiq::QueueMetrics::JobDeathMonitor.proc
     config.on(:startup) { UpgradeManager.upgrade_if_needed }
+
+    if support_death_handlers?
+      config.death_handlers << Sidekiq::QueueMetrics::JobDeathMonitor.proc
+    else
+      workers = ObjectSpace
+        .each_object(Class)
+        .select { |c| c.included_modules.include?(Sidekiq::Worker) && !c.to_s.start_with?('Sidekiq::') }
```

Comment:
@ivanetchart How are we sure that all the worker classes will be in ObjectSpace when the gem is getting initialized?
I tried it in my current project and that is not the case, the worker classes are not available at this point. But I am using sidekiq > 5.1, I am not sure if that is the problem.

Reply:
if you are using > 5 that code won't be executed because those versions support the global death handlers.
But you have a point, and yes, there may be initialization issues, if workers are loaded after this step. 

Alternatives I could think:

1. Tell users to manually include the retries_exhausted hook on each worker
2. Open the Sidekiq::Workers module and see if there's a place where we can do this for the users. This will be very tied to each sidekiq version. 
Or any other option you may have?

Reply:
I am not sure if it's feasible but can we ask the gem user for the path of the parent packages where sidekiq workers are present and load them when the gem initializes?

Reply:
Also, we can look at the Sidekiq module for other options to see if we can find a hook where we can put this code.

Reply:
it could be an option to use the `startup` lifecycle event. Not 100% sure that all workers would be loaded at that time, but it makes sense to assume that they will be. And if there are workers not loaded I would say that's probably a user fault.

Reply:
As you mentioned that if we do it on startup event we won't be sure that all the workers are loaded.

I was just looking at the sidekiq code, how about adding a sidekiq middleware which will check whether the retries has exhausted and increase the failure count. Something similar to [this]( https://github.com/mperham/sidekiq/blob/4.x/lib/sidekiq/middleware/server/retry_jobs.rb).

Reply:
Sounds like a way to go, I will give it a try. I think we should also increment the count if no retries are setup but the job also fails, right ?

Reply:
Yes, if the number of retries is 0 then we should increment the count on job failure.

Reply:
@ajitsing I went ahead and followed the suggestion. When Sidekiq version is less than 5.1, we add a new middleware to the server chain that will look for dead jobs. A job is considered dead if the retry_count is equal to 0. 

I did some local tests with sidekiq 4.2.10 and I could confirm that the failed count increased.